### ISSUE TITLE
predicting with old model(trained in spark 2.3.1, xgboost4j 0.82) in spark 3.1.1 + xgboost4j 1.5.2 error

### ISSUE THREAD

**Issue by lcx517** [dmlc/xgboost#7743]
Hi,
Because of the community support factor, we had to upgrade our spark from 2.x to 3.x.  Currently we meet a problem that there are many models trained in spark 2.3.1 + xgboost4j 0.8.2. When I predict with this model in spark 3.1.1 + xgboost4j 1.5.2, I got errors as below which I don't understand. 

Please help to solve it... Thank you~

```
22/03/22 17:01:13 ERROR scheduler.TaskSetManager: Task 33 in stage 11.0 failed 4 times; aborting job
22/03/22 17:01:13 ERROR datasources.FileFormatWriter: Aborting job 576d840c-6086-4a0b-879f-d40199321625.
org.apache.spark.SparkException: Job aborted due to stage failure: Task 33 in stage 11.0 failed 4 times, most recent failure: Lost task 33.3 in stage 11.0 (TID 127) (node62-67-11-bjzyx.qiyi.hadoop executor 5): ml.dmlc.xgboost4j.java.XGBoostError: [17:01:13] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:162: [17:01:13] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:81: Check failed: jenv->ExceptionOccurred(): 
Stack trace:
  [bt] (0) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xf883d) [0x7f599772783d]
  [bt] (1) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGBoost4jCallbackDataIterNext+0x96f) [0x7f59977286ef]
  [bt] (2) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::data::SimpleDMatrix::SimpleDMatrix<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int)+0x334) [0x7f59977fb9c4]
  [bt] (3) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::DMatrix* xgboost::DMatrix::Create<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int, std::string const&)+0x35) [0x7f59977c79e5]
  [bt] (4) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGDMatrixCreateFromDataIter+0x13f) [0x7f599773878f]
  [bt] (5) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFromDataIter+0x96) [0x7f59977262a6]
  [bt] (6) [0x7f59b9017a34]


Stack trace:
  [bt] (0) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xf883d) [0x7f599772783d]
  [bt] (1) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xdd6cb) [0x7f599770c6cb]
  [bt] (2) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::data::SimpleDMatrix::SimpleDMatrix<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int)+0x334) [0x7f59977fb9c4]
  [bt] (3) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::DMatrix* xgboost::DMatrix::Create<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int, std::string const&)+0x35) [0x7f59977c79e5]
  [bt] (4) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGDMatrixCreateFromDataIter+0x13f) [0x7f599773878f]
  [bt] (5) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFromDataIter+0x96) [0x7f59977262a6]
  [bt] (6) [0x7f59b9017a34]


    at ml.dmlc.xgboost4j.java.XGBoostJNI.checkCall(XGBoostJNI.java:48)
    at ml.dmlc.xgboost4j.java.DMatrix.<init>(DMatrix.java:54)
    at ml.dmlc.xgboost4j.scala.DMatrix.<init>(DMatrix.scala:42)
    at ml.dmlc.xgboost4j.scala.spark.XGBoostClassificationModel$$anon$1.$anonfun$batchIterImpl$1(XGBoostClassifier.scala:333)
    at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
    at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:492)
    at ml.dmlc.xgboost4j.scala.spark.XGBoostClassificationModel$$anon$1.hasNext(XGBoostClassifier.scala:345)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
    at org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer$$anon$1.hasNext(InMemoryRelation.scala:118)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
    at org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)
    at org.apache.spark.storage.memory.MemoryStore.putIteratorAsValues(MemoryStore.scala:299)
    at org.apache.spark.storage.BlockManager.$anonfun$doPutIterator$1(BlockManager.scala:1423)
    at org.apache.spark.storage.BlockManager.org$apache$spark$storage$BlockManager$$doPut(BlockManager.scala:1350)
    at org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1414)
    at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:1237)
    at org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:384)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:335)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
    at org.apache.spark.scheduler.Task.run(Task.scala:131)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
    at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2258)
    at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2207)
    at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2206)
    at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
    at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
    at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2206)
    at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1079)
    at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1079)
    at scala.Option.foreach(Option.scala:407)
    at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1079)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2445)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2387)
    at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2376)
    at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
    at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:868)
    at org.apache.spark.SparkContext.runJob(SparkContext.scala:2202)
    at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:200)
    at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:187)
    at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:108)
    at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:106)
    at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:120)
    at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
    at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3687)
    at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
    at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
    at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
    at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
    at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3685)
    at org.apache.spark.sql.Dataset.<init>(Dataset.scala:228)
    at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
    at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
    at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:615)
    at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
    at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:610)
    at deepthoughtGearSqlTool$.inputTmpTable(deepthoughtGearSqlTool.scala:296)
    at mlPredict.predicting(mlPredict.scala:71)
    at predictNode$.predict(predictNode.scala:73)
    at predictNode$.main(predictNode.scala:35)
    at predictNode.main(predictNode.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:732)
Caused by: ml.dmlc.xgboost4j.java.XGBoostError: [17:01:13] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:162: [17:01:13] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:81: Check failed: jenv->ExceptionOccurred(): 
Stack trace:
  [bt] (0) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xf883d) [0x7f599772783d]
  [bt] (1) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGBoost4jCallbackDataIterNext+0x96f) [0x7f59977286ef]
  [bt] (2) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::data::SimpleDMatrix::SimpleDMatrix<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int)+0x334) [0x7f59977fb9c4]
  [bt] (3) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::DMatrix* xgboost::DMatrix::Create<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int, std::string const&)+0x35) [0x7f59977c79e5]
  [bt] (4) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGDMatrixCreateFromDataIter+0x13f) [0x7f599773878f]
  [bt] (5) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFromDataIter+0x96) [0x7f59977262a6]
  [bt] (6) [0x7f59b9017a34]


Stack trace:
  [bt] (0) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xf883d) [0x7f599772783d]
  [bt] (1) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(+0xdd6cb) [0x7f599770c6cb]
  [bt] (2) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::data::SimpleDMatrix::SimpleDMatrix<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int)+0x334) [0x7f59977fb9c4]
  [bt] (3) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(xgboost::DMatrix* xgboost::DMatrix::Create<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int, std::string const&)+0x35) [0x7f59977c79e5]
  [bt] (4) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(XGDMatrixCreateFromDataIter+0x13f) [0x7f599773878f]
  [bt] (5) /data2/yarn/local/usercache/online01/appcache/application_1641266895554_8425499/container_e125_1641266895554_8425499_01_000007/tmp/libxgboost4j1919761885423509701.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFromDataIter+0x96) [0x7f59977262a6]
  [bt] (6) [0x7f59b9017a34]


    at ml.dmlc.xgboost4j.java.XGBoostJNI.checkCall(XGBoostJNI.java:48)
    at ml.dmlc.xgboost4j.java.DMatrix.<init>(DMatrix.java:54)
    at ml.dmlc.xgboost4j.scala.DMatrix.<init>(DMatrix.scala:42)
    at ml.dmlc.xgboost4j.scala.spark.XGBoostClassificationModel$$anon$1.$anonfun$batchIterImpl$1(XGBoostClassifier.scala:333)
    at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:486)
    at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:492)
    at ml.dmlc.xgboost4j.scala.spark.XGBoostClassificationModel$$anon$1.hasNext(XGBoostClassifier.scala:345)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
    at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
    at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
    at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
    at org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer$$anon$1.hasNext(InMemoryRelation.scala:118)
    at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
    at org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)
    at org.apache.spark.storage.memory.MemoryStore.putIteratorAsValues(MemoryStore.scala:299)
    at org.apache.spark.storage.BlockManager.$anonfun$doPutIterator$1(BlockManager.scala:1423)
    at org.apache.spark.storage.BlockManager.org$apache$spark$storage$BlockManager$$doPut(BlockManager.scala:1350)
    at org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1414)
    at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:1237)
    at org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:384)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:335)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
    at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
    at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
    at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
    at org.apache.spark.scheduler.Task.run(Task.scala:131)
    at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
    at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
    at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by trivialfis** [comment]
Hmm, the batched returned from JVM is null but there's no exception?

Could you reproduce the issue with a freshly trained model?

**Comment by lcx517** [comment]
>  the batched returned from JVM is null but there's no exception?

There's only "exception" but I don't think it's useful...
```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 9.0 failed 4 times, most recent failure: Lost task 0.3 in stage 9.0 (TID 14) (node19-72-3-bdyf.qiyi.hadoop executor 2): ml.dmlc.xgboost4j.java.XGBoostError: [17:31:07] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:162: [17:31:07] /workspace/jvm-packages/xgboost4j/src/native/xgboost4j.cpp:81: Check failed: jenv->ExceptionOccurred(): 
Stack trace:
  [bt] (0) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(+0xf883d) [0x7f58edcf583d]
  [bt] (1) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(XGBoost4jCallbackDataIterNext+0x96f) [0x7f58edcf66ef]
  [bt] (2) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(xgboost::data::SimpleDMatrix::SimpleDMatrix<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int)+0x334) [0x7f58eddc99c4]
  [bt] (3) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(xgboost::DMatrix* xgboost::DMatrix::Create<xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR> >(xgboost::data::IteratorAdapter<void*, int (void*, int (*)(void*, XGBoostBatchCSR), void*), XGBoostBatchCSR>*, float, int, std::string const&)+0x35) [0x7f58edd959e5]
  [bt] (4) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(XGDMatrixCreateFromDataIter+0x13f) [0x7f58edd0678f]
  [bt] (5) /data6/yarn/local/usercache/longyuan_antispam/appcache/application_1643652342019_13313877/container_e439_1643652342019_13313877_01_000003/tmp/libxgboost4j8013889514340339809.so(Java_ml_dmlc_xgboost4j_java_XGBoostJNI_XGDMatrixCreateFromDataIter+0x96) [0x7f58edcf42a6]
  [bt] (6) [0x7f590d017a34]
```

**Comment by lcx517** [comment]
> Could you reproduce the issue with a freshly trained model?

I'm afread not... I try to train model in spark 3.1.1, but I got error in [this issue](https://github.com/dmlc/xgboost/issues/7700)

**Comment by lcx517** [comment]
You mentioned the error from JVM.  Is it possible that the failure is caused by java version? (java sdk 1.8 + scala 2.12)

**Comment by FelixDuong** [comment]
I have found:
```..
Java 8 prior to version 8u92 support is deprecated as of Spark 3.0.0. For the Scala API, Spark 3.1.1 uses Scala 2.12. You will need to use a compatible Scala version (2.12.x)... (https://spark.apache.org/docs/3.1.1/)
```
You're running Spark 3.1.1 , so best JAVA version 11 +  ( I am using JAVA 11 ). Just my ideas !!

**Comment by lcx517** [comment]
> I have found:
> 
> ```
> Java 8 prior to version 8u92 support is deprecated as of Spark 3.0.0. For the Scala API, Spark 3.1.1 uses Scala 2.12. You will need to use a compatible Scala version (2.12.x)... (https://spark.apache.org/docs/3.1.1/)
> ```
> 
> You're running Spark 3.1.1 , so best JAVA version 11 + ( I am using JAVA 11 ). Just my ideas !!

Thank you for the information !

I have rebuilt project in Java 12 + scala 2.12.15 and run again:
1. train model in mac locally: success.
2. train model in clusters: same error as [this issue](https://github.com/dmlc/xgboost/issues/7700)
3. predict model trained freshly in local enviroment: success.
4. predict model trained in spark 2.3.1: failed with trackback as before

I tried to find some key difference between the models trained by spark 2.3.1 and spark 3.1.1.
I found in spark 2.3.1, I use `setMissing(-99999)` for model properties, which is incomatible if not setting `setAllowNonZeroForMissing(true)` in spark 3.1.1, because `allowNonZeroForMissing` is not supported in spark 2.3.1.

Now it works after I set `allowNonZeroForMissing=true`.
